File: backend/app/schemas/partiya.py
from datetime import date, datetime

from pydantic import BaseModel

from app.models.partiya import PartiyaHolati


class PartiyaOchish(BaseModel):
    mahsulot_kodi: str
    partiya_raqami: int


class PartiyaOlchov(BaseModel):
    """Faqat sort/og'irlik o'lchovlari — Tayyor mahsulotlar bo'limi ham
    to'ldira oladi. Xaridor/dogovor/narx kabi shartnoma maydonlari bu yerda
    yo'q — ularni faqat Admin PartiyaSotish orqali kiritadi."""

    sort: str | None = None
    urama_bilan_vazn: float
    urama_vazni: float
    sof_vazn: float
    kondicion_vazni: float | None = None


class PartiyaSotish(BaseModel):
    sotuv_sanasi: date
    xaridor: str
    dogovor_raqami: str | None = None
    sort: str | None = None
    urama_bilan_vazn: float
    urama_vazni: float
    sof_vazn: float
    kondicion_vazni: float | None = None
    sotuv_narxi: float | None = None


class PartiyaJavob(BaseModel):
    id: int
    mahsulot_id: int
    mahsulot_kodi: str
    mahsulot_nomi: str
    partiya_raqami: int
    holati: PartiyaHolati
    yaratilgan_vaqt: datetime
    yopilgan_vaqt: datetime | None
    kip_soni: int
    jami_kg: float

    # Sotuv ma'lumotlari (faqat "sotilgan" holatda to'ldirilgan)
    sotuv_sanasi: date | None = None
    xaridor: str | None = None
    dogovor_raqami: str | None = None
    sort: str | None = None
    urama_bilan_vazn: float | None = None
    urama_vazni: float | None = None
    sof_vazn: float | None = None
    kondicion_vazni: float | None = None
    sotuv_narxi: float | None = None
    nakladnoy_raqami: str | None = None
    nakladnoy_pdf_yoli: str | None = None

File: frontend/lib/models/partiya.dart
class Partiya {
  final int id;
  final int mahsulotId;
  final String mahsulotKodi;
  final String mahsulotNomi;
  final int partiyaRaqami;
  final String holati; // ochiq | yopiq | sotilgan
  final DateTime yaratilganVaqt;
  final DateTime? yopilganVaqt;
  final int kipSoni;
  final double jamiKg;
  final DateTime? sotuvSanasi;
  final String? xaridor;
  final String? nakladnoyRaqami;
  final String? sort;
  final double? uramaBilanVazn;
  final double? uramaVazni;
  final double? sofVazn;
  final double? kondicionVazni;

  Partiya({
    required this.id,
    required this.mahsulotId,
    required this.mahsulotKodi,
    required this.mahsulotNomi,
    required this.partiyaRaqami,
    required this.holati,
    required this.yaratilganVaqt,
    this.yopilganVaqt,
    required this.kipSoni,
    required this.jamiKg,
    this.sotuvSanasi,
    this.xaridor,
    this.nakladnoyRaqami,
    this.sort,
    this.uramaBilanVazn,
    this.uramaVazni,
    this.sofVazn,
    this.kondicionVazni,
  });

  factory Partiya.fromJson(Map<String, dynamic> j) => Partiya(
    id: j['id'],
    mahsulotId: j['mahsulot_id'],
    mahsulotKodi: j['mahsulot_kodi'],
    mahsulotNomi: j['mahsulot_nomi'],
    partiyaRaqami: j['partiya_raqami'],
    holati: j['holati'],
    yaratilganVaqt: DateTime.parse(j['yaratilgan_vaqt']),
    yopilganVaqt: j['yopilgan_vaqt'] == null
        ? null
        : DateTime.parse(j['yopilgan_vaqt']),
    kipSoni: j['kip_soni'],
    jamiKg: (j['jami_kg'] as num).toDouble(),
    sotuvSanasi: j['sotuv_sanasi'] == null ? null : DateTime.parse(j['sotuv_sanasi']),
    xaridor: j['xaridor'],
    nakladnoyRaqami: j['nakladnoy_raqami'],
    sort: j['sort'],
    uramaBilanVazn: (j['urama_bilan_vazn'] as num?)?.toDouble(),
    uramaVazni: (j['urama_vazni'] as num?)?.toDouble(),
    sofVazn: (j['sof_vazn'] as num?)?.toDouble(),
    kondicionVazni: (j['kondicion_vazni'] as num?)?.toDouble(),
  );
}
